_posts/2017-12-08-totw-131.md

```diff
@@ -0,0 +1,135 @@
+
+## What Does `=default` Do, and Why Would We Use It?
+
+Writing `=default` is our way to tell the compiler "do what you would normally
+have done for this special member function". Why might we want to do this
+rather than writing an implementation by hand or leaving the compiler to
+declare one for us?
+
+*   We can change the access level (e.g., make a constructor `protected` instead
+    of `public`), make a destructor virtual, or reinstate a function that would
+    be suppressed (e.g., a default constructor for a class that has other
+    user-declared constructors) and still get the compiler to generate the
+    function for us;
+*   Compiler-defined copy and move operations don't need maintenance every time
+    members are added or removed, if copying/moving the members is sufficient;
+*   Compiler-provided special member functions can be *trivial* (when all of the
+    operations they invoke are themselves trivial), which can make them faster
+    and safer;
+*   Types with defaulted constructors can be *aggregates*, and hence support
+    *aggregate initialization*, whereas those with user-provided constructors
+    cannot.
+*   Explicitly declaring a defaulted member gives us a place to document the
+    semantics of the resulting function;
+*   In a class template, `=default` gives a simple way to conditionally provide
+    an operation depending on whether some underlying type provides it.
+
+When we use `=default` on the initial declaration of a special member function,
+the compiler will check whether it can synthesize an inline definition for that
+function. If it can, it does. If it can't, the function is actually declared as
+*deleted*, just as if we'd written `=delete`. That's just what we'd need for
+transparently wrapping a class (say, if we're defining a class template), but it
+can be surprising to readers.
+
+If a function's initial declaration uses `=default`, or if the compiler declares
+a special member function that is not user-declared, an appropriate `noexcept`
+specification is deduced, potentially allowing faster code.
+
+## How Does It Work?
+
+Before C++11, if we needed a default constructor and already had other
+constructors, we'd have written one as:
+
+```c++
+class A {
+ public:
+  A() {}  // User-provided, non-trivial constructor makes A a non-aggregate.
+};
+```
+
+From C++11 onwards we have more options.
+
+```c++
+class C {
+ public:
+  C() = default;  // misleading: C has a deleted default constructor
+ private:
+  const int i;  // const => must always be initialized.
+};
+
+class D {
+ public:
+  D() = default;  // unsurprising, but not explicit: D has a default constructor
+ private:
+  std::unique_ptr<int> p;  // std::unique_ptr has a default constructor
+};
+```
+
+Clearly we shouldn't write code like `class C`: in a non-template, use
+`=default` only if you intend the class to support the operation (and then test
+that it does).  clang-tidy includes a check for this.
```

Comment:
bold and code clang-tidy? (That's what I usually do for tools/command-lines.)

*`clang-tidy*`

Reply:
Done.